Src/IronPython/Runtime/PythonEncoding.cs

```diff
@@ -614,43 +626,132 @@ public override bool Fallback(char charUnknownHigh, char charUnknownLow, int ind
+                => GetCharCount(bytes, index, count, flush: true);
+
+            public override int GetCharCount(byte[] bytes, int index, int count, bool flush)
+                => _decoder.GetCharCount(bytes, index, count, flush);
+
+            public override unsafe int GetCharCount(byte* bytes, int count, bool flush)
+                => _decoder.GetCharCount(bytes, count, flush);
+
+            public override int GetChars(byte[] bytes, int byteIndex, int byteCount, char[] chars, int charIndex)
+                => GetChars(bytes, byteIndex, byteCount, chars, charIndex, flush: true);
+
+            public override int GetChars(byte[] bytes, int byteIndex, int byteCount, char[] chars, int charIndex, bool flush)
+                => _decoder.GetChars(bytes, byteIndex, byteCount, chars, charIndex, flush);
+
+            public override unsafe int GetChars(byte* bytes, int byteCount, char* chars, int charCount, bool flush)
+                => _decoder.GetChars(bytes, byteCount, chars, charCount, flush);
+
+            public override void Reset() => _decoder.Reset();
+
+            private class ProxyDecoderFallback : DecoderFallback {
+                private readonly DecoderFallbackBuffer _buffer;
+                private readonly int _maxCharCount;
+
+                public ProxyDecoderFallback(DecoderFallbackBuffer buffer, int maxCharCount) {
+                    _buffer = buffer;
+                    _maxCharCount = maxCharCount;
+                }
+
+                public override DecoderFallbackBuffer CreateFallbackBuffer() => _buffer;
+                public override int MaxCharCount => _maxCharCount;
+            }
+        }
+
         private class PythonDecoder : Decoder {
             private readonly PythonEncoding _parentEncoding;
             private readonly Decoder _pass1decoder;
             private Decoder? _pass2decoder;
 
             public PythonDecoder(PythonEncoding parentEncoding) {
                 _parentEncoding = parentEncoding;
-                _pass1decoder = _parentEncoding.Pass1Encoding.GetDecoder();
+                _pass1decoder = GetDecoder(_parentEncoding.Pass1Encoding);
+            }
+
+            private static Decoder GetDecoder(Encoding encoding) {
+                Decoder decoder = encoding.GetDecoder();
+
+                if (!(decoder.Fallback is PythonDecoderFallback) && encoding.DecoderFallback is PythonDecoderFallback) {
+                    // Non-conformant Decoder implementation, the challenge is to get to the fallback buffer used by such decoder.
+                    // See notes at PythonEncoder.GetEncoder(...)
+                    if (decoder.GetType().FullName == "System.Text.DecoderNLS") return decoder;
+                    decoder = new ProxyDecoder(encoding);
+                }
+                return decoder;
             }
 
+            private static PythonDecoderFallbackBuffer? GetPythonDecoderFallbackBuffer(Decoder? enc) {
```

Comment:
Good catch. It is a bug. I overlooked it because it the standard `DecoderNLS` will never generate errors (as far as I know) so I don't have any tests for such case. And a subclassed decoder (like `ISO2022Decoder`) is being (wrongly) wrapped in a `ProxyDecoder`, which makes it work for simple cases but would break down during incremental decoding (a code path not used by IronPython yet).